providers/implementations/signature/dsa_sig.c

```diff
@@ -381,94 +547,95 @@ static int dsa_digest_sign_init(void *vpdsactx, const char *mdname,
+    /* Sigalg implementations shouldn't do digest_sign */
+    if (pdsactx->flag_sigalg)
         return 0;
 
-    /*
-     * If sig is NULL then we're just finding out the sig size. Other fields
-     * are ignored. Defer to dsa_sign.
-     */
-    if (sig != NULL) {
-        /*
-         * There is the possibility that some externally provided
-         * digests exceed EVP_MAX_MD_SIZE. We should probably handle that somehow -
-         * but that problem is much larger than just in DSA.
-         */
-        if (!EVP_DigestFinal_ex(pdsactx->mdctx, digest, &dlen))
-            return 0;
-    }
+    ok = dsa_sign_message_final(pdsactx, sig, siglen, sigsize);
 
     pdsactx->flag_allow_md = 1;
 
-    return dsa_sign(vpdsactx, sig, siglen, sigsize, digest, (size_t)dlen);
+    return ok;
 }
 
+static int dsa_digest_verify_init(void *vpdsactx, const char *mdname,
+                                  void *vdsa, const OSSL_PARAM params[])
+{
+    return dsa_digest_signverify_init(vpdsactx, mdname, vdsa, params,
+                                      EVP_PKEY_OP_VERIFYMSG,
+                                      "DSA Digest Verify Init");
+}
+
+static int dsa_digest_verify_update(void *vpdsactx, const unsigned char *data,
+                                    size_t datalen)
+{
+    PROV_DSA_CTX *pdsactx = (PROV_DSA_CTX *)vpdsactx;
+
+    if (pdsactx == NULL)
+        return 0;
+    /* Sigalg implementations shouldn't do digest_sign */
+    if (pdsactx->flag_sigalg)
+        return 0;
+
+    return dsa_signverify_message_update(vpdsactx, data, datalen);
+}
 
 int dsa_digest_verify_final(void *vpdsactx, const unsigned char *sig,
                             size_t siglen)
 {
     PROV_DSA_CTX *pdsactx = (PROV_DSA_CTX *)vpdsactx;
-    unsigned char digest[EVP_MAX_MD_SIZE];
-    unsigned int dlen = 0;
+    int ok = 0;
 
-    if (!ossl_prov_is_running() || pdsactx == NULL || pdsactx->mdctx == NULL)
+    if (pdsactx == NULL)
         return 0;
-
-    /*
-     * There is the possibility that some externally provided
-     * digests exceed EVP_MAX_MD_SIZE. We should probably handle that somehow -
-     * but that problem is much larger than just in DSA.
-     */
-    if (!EVP_DigestFinal_ex(pdsactx->mdctx, digest, &dlen))
+    /* Sigalg implementations shouldn't do digest_verify */
+    if (pdsactx->flag_sigalg)
         return 0;
 
+    if (dsa_verify_set_sig(pdsactx, sig, siglen))
+        ok = dsa_verify_message_final(vpdsactx);
+
     pdsactx->flag_allow_md = 1;
 
-    return dsa_verify(vpdsactx, sig, siglen, digest, (size_t)dlen);
+    return ok;
 }
 
 static void dsa_freectx(void *vpdsactx)
 {
     PROV_DSA_CTX *ctx = (PROV_DSA_CTX *)vpdsactx;
 
-    OPENSSL_free(ctx->propq);
     EVP_MD_CTX_free(ctx->mdctx);
     EVP_MD_free(ctx->md);
-    ctx->propq = NULL;
-    ctx->mdctx = NULL;
-    ctx->md = NULL;
+    OPENSSL_free(ctx->sig);
+    OPENSSL_free(ctx->propq);
     DSA_free(ctx->dsa);
-    OPENSSL_free(ctx);
+    OPENSSL_clear_free(ctx, sizeof(*ctx));
```

Comment:
Why was this changed?

Reply:
'cause it's there in rsa_sig.c, so it looked to me like it was omitted by mistake here.  Like, why have that in one signature implementation and not the others?

Reply:
Not sure why that one does it either. Unless there is something sensitive in there.

Reply:
I'm struggling to see what is sensitive in the `PROV_DSA_CTX` structure that requires this. Although, TBH, I'm struggling with the same question for `PROV_RSA_CTX`. So perhaps it's the other way around....RSA does this in error??

Reply:
À look through git history may show why it was done...

Reply:
Hmmm, nope, no git history to lean on.  `OPENSSL_clear_free()` was there from the start in the PR where this implementation was first made, see https://github.com/openssl/openssl/pull/10557/files#diff-c7e8f595b78166fd37d007bb71d5f58e01fe4602b6a319ad790beaaeeee84439R683-R697

So I guess it's lost in a conversation we must have had at the time, or I dunno what.

Reply:
Without a clear need I would remove this change from this PR. Just because RSA does it doesn't seem like a strong enough reason - since the reason that RSA does it looks suspect.

Reply:
Okie